dask/highlevelgraph.py

```diff
@@ -24,6 +27,122 @@ def _find_layer_containing_key(key):
+    """
+
+    def cull(self, keys: Set) -> "Layer":
+        """Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
+        >>> d.cull({'out'})  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+
+        seen = set()
+        out = {}
+        work = set(flatten(keys))
+        while work:
+            new_work = set()
+            for k in work:
+                out[k] = self[k]
+                for d in get_dependencies(self, k):
+                    if d not in seen:
+                        seen.add(d)
+                        new_work.add(d)
+            work = new_work
+
+        return BasicLayer(out)
+
+    def get_external_dependencies(self, all_hlg_keys: Set) -> Set:
+        """Get external dependencies
+
+        Parameters
+        ----------
+        all_hlg_keys : set
+            Set of all keys in the high level graph.
+
+        Returns
+        -------
+        deps: set
+            Set of dependencies
+        """
+
+        all_deps = keys_in_tasks(all_hlg_keys, self.values())
+        return all_deps.difference(self.keys())
+
+
+class BasicLayer(Layer):
+    """Basic implementation of `Layer`
+
+    Parameters
+    ----------
+    mapping : Mapping
+        The mapping between keys and tasks, typically a dask graph.
+    key_dependencies : Mapping[Hashable, Set], optional
+        Mapping between keys and their dependencies
+    unstructured_tasks : Mapping[Hashable, Set], optional
+        Mapping between keys and unstructured tasks, which are tasks
+        that have no dependencies information.
+    """
+
+    def __init__(self, mapping, key_dependencies=None, unstructured_tasks=None):
+        self._mapping = mapping
+        self._key_dependencies = key_dependencies
+        self._unstructured_tasks = unstructured_tasks
+
+    def __getitem__(self, k):
+        return self._mapping[k]
+
+    def __iter__(self):
+        return iter(self._mapping)
+
+    def __len__(self):
+        return len(self._mapping)
+
+    def get_external_dependencies(self, all_hlg_keys):
+        if self._key_dependencies is None or self._unstructured_tasks is None:
+            return super().get_external_dependencies(all_hlg_keys)
+
+        ret = set()
+        for v in self._key_dependencies.values():
+            ret.update(v)
+        ret |= keys_in_tasks(all_hlg_keys, self._unstructured_tasks.values())
+        return ret
+
+    def cull(self, keys):
+        if self._key_dependencies is None or self._unstructured_tasks is None:
+            return super().cull(keys)
+
+        # TODO: Are we sure that tasks in `self` never depend on
+        # other tasks in `self`?
```

Comment:
It is common for tasks in one layer to depend on other tasks in that same layer.  Consider for example sequential rolling operations.

Reply:
Fixed

Reply:
Fixed and added a test of this case.